rfc/system/3749-sane-overridable-defaults.md

```diff
@@ -0,0 +1,208 @@
+* project defaults
+* per-workflow defaults
+
+
+#### Proposed document structure
+A single call to get all overrides should return the collated document for all overrides set at a level (e.g. per project)
+
+```
+$ flytectl get overrides -p flytesnacks -d development --workflow MyExampleWorkflow
+
+apiVersion: 1  # Version of this document structure
+domain: development
+project: flytesnacks
+workflow: MyExampleWorkflow
+kind: overrides
+spec:
+  task_resources:
+    defaults:
+      cpu: "1"
+      memory: "500Mi"
+    limits:
+      cpu: "3"
+      memory: "2Gi"
+      gpu: "4"
+  workflow_execution_config:
+    inherited: true
+    source: project
+    labels:
+      my-key: "my-value"
+    security_context:
+      core_identity:
+        iam_role: "my_iam_role"
+status:
+  version: 1 # Version for the entire overrides document
+```
+
+Note a few details
+* The spec is the yaml formatted list of all available [matchable resource messages](https://github.com/flyteorg/flyteidl/blob/master/protos/flyteidl/admin/matchable_resource.proto) defined
+* Specific matchable resources can be inherited, that is, not set at the workflow level used to get the document. In this example, the workflow execution config is set at the project level
+* The status includes a version. This is managed on the backend.
+
+
+### Implementation Proposal
+Define a new set of APIs for returning overridable defaults
+
+
+#### Read Experience
+```
+rpc GetOverrides(GetOverridesRequest) returns (GetOverridesResponse)
+
+...
+
+message OverridesDocument {
+  string domain = 1;
+  
+  string project = 2; // optional
+  
+  string workflow = 3; // optional
+  
+  int32 api_version = 4;
+  
+  repeated ResourceOverrides overrides = 5;
+  
+  OverridesStatus status = 6;
+}
+
+message ResourceOverrides {
+  MatchingAttributes attributes = 1; # Oneof for existing matchable resources
+  
+  bool inherited = 2;
+  
+  string domain = 3;  // optional, if inherited
+  
+  string project = 4;  // optional, if inherited
+  
+  ...
+}
+
+message OverridesStatus {
+  int32 version = 1;
+}
+
+```
+
+##### Write Experience
+
+Traditionally overrides are managed programmatically through `flytectl` however declarative, [GitOps](https://about.gitlab.com/topics/gitops/) style management can be very useful for managing infrastructure as code. To that end, this proposal includes using a static config map for managing configurable overrides as the source of truth.
+
+A migration strategy defined below outlines how existing users of overridable resources who have built up overrides in the database using flytectl can cut-over to the declarative style. FlyteAdmin deployments can be made configurable, so that overrides are either read from the existing database (for backwards compatibility with the current implementation) or a configmap.
+
+In FlyteAdmin config
+
+```
+flyteadmin:
+  overrides:
+    # Points to deprecated matchable overrides written by flytectl
```

Comment:
> # Points to deprecated matchable overrides written by flytectl

Does this mean that creating overrides with flytectl will be deprecated alltogether?